File: src/ios/PKVideoThumbnail.m
/**
 * PKVideoThumbnail
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies
 * or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
 * OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#import "PKVideoThumbnail.h"
#import <Cordova/CDVPluginResult.h>
#import <AVFoundation/AVFoundation.h>
#import <AVFoundation/AVAsset.h>
#import "CDVFile.h"
#import <MediaPlayer/MediaPlayer.h>

@implementation PKVideoThumbnail

- (NSURL *) obtainURLForPath:(NSString *)path {
    if ([path hasPrefix:@"cdvfile://"]) {
        // use the File API to get the appropriate URL, given the path
        // based on media plugin's code for obtaining file paths
        CDVFile *filePlugin = [self.commandDelegate getCommandInstance:@"File"];
        CDVFilesystemURL *fsURL = [CDVFilesystemURL fileSystemURLWithString:path];
        NSString *filePath = [filePlugin filesystemPathForURL:fsURL];
        if (filePath) {
            return [NSURL URLWithString:[[@"file://" stringByAppendingString:filePath] stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
        } else {
            return [NSURL URLWithString:[path stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
        }
    } else {
        if ([path rangeOfString:@"://"].location == NSNotFound) {
            NSString *pathForResource = [self.commandDelegate pathForResource:path];
            if (pathForResource) {
                return [NSURL URLWithString:[[@"file://" stringByAppendingString:pathForResource] stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
            } else {
                return NULL;
            }
        }
        else {
            return [NSURL URLWithString:[path stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
        }
    }
}

- (CDVPluginResult *) extractThumbnailAtPath:(NSString *)sourcePath toPath:(NSString *)targetPath withOptions:(NSDictionary *)options
{

    UIImage *thumbnail;
    NSURL *url = [self obtainURLForPath:sourcePath];

    // from http://stackoverflow.com/q/9145968 by Mx Gherkins
    // and http://www.catehuston.com/blog/2015/07/29/ios-getting-a-thumbnail-for-a-video/

    AVAsset *asset = [AVAsset assetWithURL:url];
    //AVURLAsset *asset = [[AVURLAsset alloc] initWithURL:url options:nil];
    AVAssetImageGenerator *generate = [AVAssetImageGenerator assetImageGeneratorWithAsset:asset];
    //AVAssetImageGenerator *generate = [[AVAssetImageGenerator alloc] initWithAsset:asset];
    generate.appliesPreferredTrackTransform = YES; // http://stackoverflow.com/a/9146246 by djromero

    NSError *err = NULL;
    Float64 position = [[options objectForKey:@"position"] floatValue];
    CMTime time = CMTimeMakeWithSeconds(position, 1000);
    CGImageRef imgRef = [generate copyCGImageAtTime:time actualTime:NULL error:&err];
    if (err) {
        return [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Could not extract thumbnail from %@ at time %f; err=%@", url, position, err]];
    }
    thumbnail = [[UIImage alloc] initWithCGImage:imgRef];
    CGImageRelease(imgRef);

    if (!thumbnail) {
        return [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Didn't get a thumbnail from CGImage using %@", url]];
    }

    // if we have a thumbnail, we now need to determine if it needs resized...
    NSDictionary *dimensions = [options objectForKey:@"resize"];
    if (dimensions) {
        // from http://nshipster.com/image-resizing/
        Float64 newWidth = [[dimensions objectForKey:@"width"] floatValue];
        Float64 newHeight = [[dimensions objectForKey:@"height"] floatValue];
        CGRect targetFrame = CGRectMake(0, 0, newWidth, newHeight);
        CGRect targetFrameWithAspectRatio = AVMakeRectWithAspectRatioInsideRect(thumbnail.size, targetFrame);

        UIGraphicsBeginImageContextWithOptions(targetFrameWithAspectRatio.size, true, 0.0);
        [thumbnail drawInRect:CGRectMake(0.0, 0.0,
                                         targetFrameWithAspectRatio.size.width,
                                         targetFrameWithAspectRatio.size.height)];
        thumbnail = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
    }

    Float64 quality = [[options objectForKey:@"quality"] floatValue];

    // if mode is file, we'll write it out to a file as a JPEG
    if ([[[options objectForKey:@"mode"] lowercaseString] isEqualToString:@"file"]) {
        NSURL *target = [self obtainURLForPath:targetPath];
        if (target) {
            NSString *revisedTargetPath = [target.absoluteString stringByReplacingOccurrencesOfString:@"file://" withString:@""];
            // write out the thumbnail; a return of NO will be a failure.
            if([UIImageJPEGRepresentation(thumbnail, quality) writeToFile:revisedTargetPath atomically:YES]) {
            return [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[NSString stringWithFormat:@"%@", targetPath]];
            } else {
                return [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Could not create thumbnail. Source: %@, Target: %@, Modified Target: %@", url, targetPath, revisedTargetPath]];
            };
        } else {
            return [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Target can't be in the bundle. Source: %@, Target: %@", url, targetPath]];
        }
    } else {
        // return the image data to the callback
        NSData *imageData = UIImageJPEGRepresentation(thumbnail, quality);
        if ([[[options objectForKey:@"mode"] lowercaseString] isEqualToString:@"array"]) {
            return [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArrayBuffer:imageData];
        } else {
            return [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[@"data:image/jpeg;base64," stringByAppendingString:[imageData base64EncodedStringWithOptions:0]]];
        }
    }
}

- (void) createThumbnail:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    @try {
        NSString* theSourceVideoName = [command argumentAtIndex:0];
        NSString* theTargetImageName = [command argumentAtIndex:1];

        NSMutableDictionary* options = nil;
        options = [command argumentAtIndex:2 withDefault:nil];
        if (options != nil) {
            options = [NSMutableDictionary dictionaryWithDictionary:options];
        } else {
            options = [NSMutableDictionary dictionaryWithCapacity:1];
        }

        // supply default options if none provided
        if (![options objectForKey:@"position"]) {
            // if not supplied, we default to 1 second into the video
            [options setObject:[NSNumber numberWithFloat:1.0] forKey:@"position"];
        }
        if (![options objectForKey:@"mode"]) {
            // if not supplied, we default to "file" mode (rather than "data")
            [options setObject:@"file" forKey: @"mode"];
        }
        if (![options objectForKey:@"quality"]) {
            // if not supplied, thumbnail quality is 80%.
            [options setObject:[NSNumber numberWithFloat:0.8] forKey:@"quality"];
        }

        pluginResult = [self extractThumbnailAtPath:theSourceVideoName toPath:theTargetImageName withOptions:options];
    } @catch (NSException* exception) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_JSON_EXCEPTION messageAsString:[exception reason]];
    }

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}
@end

File: src/android/PKVideoThumbnail.java
/**
 * PKVideoThumbnail
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following
 * conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies
 * or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
 * OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
package com.photokandy.PKVideoThumbnail;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaResourceApi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.Matrix;
import android.graphics.RectF;
import android.media.*;
import android.provider.MediaStore;
import android.util.Base64;
import android.net.Uri;

import java.io.*;

/**
 * This class echoes a string called from JavaScript.
 */
public class PKVideoThumbnail extends CordovaPlugin {
    
    private String mapPath(String path) {
        CordovaResourceApi resourceApi = webView.getResourceApi();
        String returnUri;
        try {
            returnUri = (resourceApi.remapUri(Uri.parse(path))).toString();
        } catch (IllegalArgumentException e) {
            returnUri = path;
        }
        return returnUri;
    }

    /**
     * Executes the request and returns PluginResult.
     *
     * @param action        The action to execute.
     * @param args          JSONArry of arguments for the plugin.
     * @param callbackId    The callback id used when calling back into JavaScript.
     * @return              A PluginResult object with a status and message.
     */
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        try {
            if (action.equals("createThumbnail")) {
                String sourceVideo = mapPath(args.getString(0)).replaceAll("file://", "");
                String targetImage = mapPath(args.getString(1)).replaceAll("file://", "");

                JSONObject options = args.getJSONObject(2);

                String processingMode = options.optString("mode", "file");
                int outputQuality = (int) (options.optDouble("quality", 0.8) * 100);
                double position = options.optDouble("position", 1.0);

                JSONObject resize = options.optJSONObject("resize");

                // get the thumbnail
                Bitmap thumbnail = ThumbnailUtils.createVideoThumbnail(sourceVideo, MediaStore.Images.Thumbnails.MINI_KIND);
                // NOTE: can't get the thumbnail at a specific location :-(
                if (thumbnail == null) {
                    callbackContext.error( "Couldn't read video at " + sourceVideo);
                    return true;
                }
                // resize the thumbnail if necessary
                if (resize != null) {
                    double newWidth = resize.optDouble("width", 100);
                    double newHeight = resize.optDouble("height", 100);
                    
                    // based on http://stackoverflow.com/a/30024180/741043
                    Matrix m = new Matrix();
                    m.setRectToRect(new RectF(0, 0, thumbnail.getWidth(), thumbnail.getHeight()), 
                                    new RectF(0, 0, (float)newWidth, (float)newHeight), Matrix.ScaleToFit.CENTER);
                    try {
                        thumbnail = Bitmap.createBitmap(thumbnail, 0, 0, thumbnail.getWidth(), thumbnail.getHeight(), m, true);
                    } catch (IllegalArgumentException e) {
                        callbackContext.error( "Could not resize thumbnail; width or height out of range?" );
                        return true;                                            
                    }
                }

                // compress the thumbnail to a JPEG                 
                ByteArrayOutputStream theThumbnailOS = new ByteArrayOutputStream();
                if (!thumbnail.compress(CompressFormat.JPEG, outputQuality, theThumbnailOS)) {
                    callbackContext.error( "Could not compress thumbnail to a JPEG of the desired quality." );
                    return true;                    
                }
                
                // and return it in the desired format
                if (processingMode.equals("file")) {
                    FileOutputStream theOutputStream;
                    try
                    {
                        File theOutputFile = new File (targetImage);
                        if (!theOutputFile.exists())
                        {
                            if (!theOutputFile.createNewFile())
                            {
                                callbackContext.error("Could not save thumbnail to " + targetImage);
                                return true;
                            }
                        }
                        if (theOutputFile.canWrite())
                        {
                            theOutputStream = new FileOutputStream (theOutputFile);
                            if (theOutputStream != null)
                            {
                                theThumbnailOS.writeTo(theOutputStream);
                            }
                            else
                            {
                                callbackContext.error("Could not save thumbnail; target not writeable");
                                return true;
                            }
                        }
                    }
                    catch (IOException e)
                    {
                        e.printStackTrace();
                        callbackContext.error( "I/O exception saving thumbnail" );
                        return true; 
                    }
                    callbackContext.success(targetImage);
                    return true; 
                }
                if (processingMode.equals("array")) {
                    callbackContext.success(theThumbnailOS.toByteArray());
                    return true;
                }
                // otherwise return base64
                // based on http://stackoverflow.com/a/9768973/741043
                callbackContext.success("data:image/jpeg;base64," + 
                                        Base64.encodeToString(theThumbnailOS.toByteArray(), Base64.DEFAULT));
                return true; 
            } else {
                return false;
            }
        } catch (JSONException e) {
            callbackContext.error ( "JSON Exception" );
            return true;
        }
    }
}
